testnet/advancedTokenSale.js

```diff
@@ -67,67 +67,72 @@ function sleep(ms) {
       console.log('==============================')
       console.log('Sleeping before iteration ', i)
       console.log('------------------------------')
-      await sleep(9000);
-
-      const prevLockingScript = advTokenSale.lockingScript
+      await sleep(6000);
 
       const numBought = boughtEachIteration[i]
       const numBoughtHex = num2bin(numBought, DataLen)
       const spendAmount = numBought * satsPerToken
 
-      const newState = [advTokenSale.dataPart.toASM(),  toHex(publicKeys[i]) + numBoughtHex].join(' ');
+      const newState = [advTokenSale.dataPart.toASM(), toHex(publicKeys[i]) + numBoughtHex].join(' ');
       // build-up a list of sales
-      advTokenSale.setDataPart(newState)
+      const newLockingScript = bsv.Script.fromASM([advTokenSale.codePart.toASM(), newState].join(' '))
 
       // Increase contract funding to match proceeds from sale
       // The contract expects/enforces this
       const newAmount = amount + spendAmount
 
-      let unlockingTx = await createUnlockingTx(lockingTxid, amount, prevLockingScript, newAmount, advTokenSale.lockingScript)
-
-
-      unlockingTx = await anyOnePayforTx(unlockingTx, privateKey.toAddress(), FEE)
-
-
-      const changeAmount = unlockingTx.outputs[unlockingTx.outputs.length -1].satoshis
-
-      const preimage = getPreimage(unlockingTx, prevLockingScript, amount, 0, sighashType)
-
-      const unlockingScript = advTokenSale.buy(
-        new SigHashPreimage(toHex(preimage)), // sighashPreimage
-        new Ripemd160(toHex(pkh)), // changePKH
-        changeAmount, // changeSats
-        new Bytes(toHex(publicKeys[i])), // buyer's public key
-        numBought // number of tokens purchased
-      ).toScript()
-
-      // unlock contract input
-      unlockingTx.inputs[0].setScript(unlockingScript)
-
-      // unlock other p2pkh inputs
-      for (let index = 1; index < unlockingTx.inputs.length; index++) {
-        unlockP2PKHInput(privateKey, unlockingTx, index, sighashType)
-      }
+      const unlockingTx = new bsv.Transaction();
+
+      unlockingTx.addInput(createInputFromPrevTx(prevTx))
+        .from(await fetchUtxos(privateKey.toAddress()))
+        .addOutput(new bsv.Transaction.Output({
+          script: newLockingScript,
+          satoshis: newAmount,
+        }))
+        .change(privateKey.toAddress())
+        .setInputScript(0, (tx, _) => {
+          let preimage = getPreimage(
+            tx,
+            advTokenSale.lockingScript,
+            amount,
+            0,
+            sighashType
+          );
+
+          return advTokenSale.buy(
+            new SigHashPreimage(toHex(preimage)), // sighashPreimage
+            new Ripemd160(toHex(pkh)), // changePKH
+            tx.getChangeAmount(), // changeSats
+            new Bytes(toHex(publicKeys[i])), // buyer's public key
+            numBought // number of tokens purchased
+          ).toScript();
+
+        })
+        .seal()
```

Comment:
What is this?

Reply:
seal Transaction to call callback automatically